src/Metric.jl

```diff
@@ -0,0 +1,202 @@
+@doc doc"""
+    Metric
+
+Abstract type for the pseudo-Riemannian metric tensor $g$, a family of smoothly
+    varying inner products on the tangent space. See [`inner`](@ref).
+"""
+abstract type Metric end
+
+@doc doc"""
+    RiemannianMetric <: Metric
+
+Abstract type for Riemannian metrics, a family of positive definite inner
+    products. The positive definite property means that for $v \in T_x M$,
+    the inner product $g(v,v) > 0$.
+"""
+abstract type RiemannianMetric <: Metric end
+
+"""
+    LorentzMetric <: Metric
+
+Abstract type for Lorentz metrics, which have a single temporal dimension.
+"""
+abstract type LorentzMetric <: Metric end
+
+"""
+    MetricManifold{M<:Manifold,G<:Metric} <: Manifold
+
+Equip a manifold with a metric. Such a manifold is generally called pseudo-
+or semi-Riemannian. Each `MetricManifold` must implement
+[`local_metric`](@ref).
+
+# Constructor
+
+    MetricManifold(manifold, metric)
+"""
+struct MetricManifold{M<:Manifold,G<:Metric} <: Manifold
+    manifold::M
+    metric::G
+end
+
+@traitimpl IsDecoratorManifold{MetricManifold}
+
+@traitdef IsRiemannianManifold{M}
+@traitimpl IsRiemannianManifold{M} <- isriemannianmanifold(M)
+
+isriemannianmanifold(::Type{MT}) where {MT<:Manifold} = false
```

Comment:
I know what you would like to achieve with this function, but it seems to be inconsistent, since this would lead to the case, that just `Sphere(3)` or `Euclidean(5)` are not riemannian manifolds, which is misleading. Not necessarily type-wise, but from the perception of the user, since both actually _are_ riemannian (just not _explicitly_ equipped with a metric in our toolbox at that point).

 could we do something like `hasMetric(::Type{MetricManifold{MT,GT},::GTc) where you could check whether `GT` and `GTc` are of same type? This would be more general than isriemannianmaifold – you could also check whether a `MetricManifold` has a `Lorentzian` metric – and it would just not be defined for other cases than `MetricManifolds`.

Reply:
Good point. Check out https://github.com/JuliaNLSolvers/ManifoldMuseum/pull/12/commits/5d0ed52d3edcebc531cec834e05023ac265199bb. It provides a simple interface for indicating the default metric for a manifold and getting automatic forwarding of key functions.